### PR TITLE
Adds jdbc macros from internal

### PR DESCRIPTION
**Pull request by rubanm**
Part 1 of the new `scalding-db` and `scalding-db-macros` subprojects. Addresses https://github.com/twitter/scalding/issues/1124

This adds @ianoc's original JDBC macros along with some refactoring and fixes done during internal use at Twitter.

Part 2 to be followed will be the new JDBC source and sink classes based on these macros.

---

The motivation for this is to have an improved way of defining JDBC Typed Source and Sinks using **two simple steps**:

1) Defining a case class that represents you table schema:

``` scala
case class ExampleDBRecord(
  user_id: Long,
  tweet_id: Long,
  created_at: java.util.Date,
  deleted: Boolean)
```

2) Defining a TypedJDBCSource based on the above case class:

``` scala
case class ExampleDBRecordsTable(implicit dbsInEnv : AvailableDatabases)
    extends TypedJDBCSource[ExampleDBRecord](dbsInEnv) {
  override val tableName = TableName("example_table")
  override val database = Database("example_schema") 
}
```

Under the hood, this case class is automatically mapped to the underlying DB schema using a **macro-generated** `DBTypeDescriptor`. As a user, one does not need to specify SQL mappings or Injections to do this back and forth translation.

---

There are also optimizations for avoiding common pitfalls when talking to databases directly from hadoop nodes (too many open connections, inefficient OFFSET based queries, table lock contention). This includes performing a jdbc -> hdfs snapshot via the submitter for smaller datasets and likewise for writes. These will be in a separate PR along with the new jdbc source classes.

### CODE REVIEW

File: project/plugins.sbt [hunk 1]
```diff
@@ -13,4 +13,6 @@ addSbtPlugin("com.eed3si9n" % "sbt-assembly" % "0.10.2")
 
 addSbtPlugin("com.typesafe.sbt" % "sbt-ghpages" % "0.5.1")
 
+addSbtPlugin("com.typesafe" % "sbt-mima-plugin" % "0.1.6")
```
Comment (ianoc) [inline project/plugins.sbt]: mima is already in this file

**Comment by ianoc** [comment]
@rubanm can you add in the description some examples of using the new jdbc stuff to show motivation for this stuff

**Comment by ianoc** [comment]
Looks great though, thanks for splitting this out

**Comment by rubanm** [comment]
@ianoc updated the description

**Comment by johnynek** [comment]
can we add a README or something that explains how to get going with this? It's great, but reading the code to learn it is pretty hard.

**Comment by rubanm** [comment]
@johnynek makes sense. Added a README. It probably needs a few more edits though.

File: scalding-db-macros/README.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+------------- | -------------
+`Int` | `INTEGER`
+`Long` | `BIGINT`
+`Short` | `SMALLINT`
+`Double` | `DOUBLE`
+`@varchar @size(20) String `| `VARCHAR(20)`
+`@text String` | `TEXT`
+`java.util.Date` | `DATETIME`
+`@date java.util.Date` | `DATE`
+`Boolean` | `BOOL`, `BOOLEAN`, `TINYINT`
```
Comment (johnynek) [inline scalding-db-macros/README.md]: what is up with the three values on the right in this column?

File: scalding-db-macros/README.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+* Annotations are used for String types to clearly distinguish between TEXT and VARCHAR column types
+* Scala `Option`s can be used to denote columns that are `NULLABLE` in the DB
+* Nested case classes can be used as a workaround for the 22-size limitation on Scala tuples, case classes
```
Comment (johnynek) [inline scalding-db-macros/README.md]: this seems to stop abruptly: case classes... ? what? I think you are going to explain that case classes are also flattened in left to right order. You might want to show an example of that.

File: scalding-db-macros/README.md (hunk 2 above)
Comment (johnynek) [inline scalding-db-macros/README.md]: Seems like the SQL type is BOOLEAN, from reading the code below.

File: scalding-db-macros/README.md (hunk 3 above)
Comment (rubanm) [inline scalding-db-macros/README.md]: Added a nested case classes example.

**Comment by johnynek** [comment]
Are the tests being run? Did we update the travis running script?

We really need a more reliable way to make sure we are testing our sub-modules.

**Comment by rubanm** [comment]
Updated the travis script to include the new db modules.

File: scalding-db-macros/README.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+Case class representing your DB schema:
+
+    scalding> case class ExampleDBRecord(
+       |   card_id: Long,
+       |   tweet_id: Long,
+       |   created_at: Option[java.util.Date],
+       |   deleted: Boolean = false)
+    defined class ExampleDBRecord
+
```
Comment (johnynek) [inline scalding-db-macros/README.md]: what's the last step? How to I read or write from a Database? Seems like we are stopping short of clearly explaining the use of the whole package here.

Comment (rubanm) [inline scalding-db-macros/README.md]: Yes, the sources will be added in a follow-up PR. So this is a little incomplete in that sense.

**Pull request merged**